[autograd][c++ API] bind autograd functions into C++

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #23913 [jit] bind autograd.backward and tensor.backward in TorchScript
* #24871 [jit] bind autograd.grad function into TorchScript
* **#24342 [autograd][c++ API] bind autograd functions into C++**

Summary:

Right now the two APIs that provided in autograd package only have
python bindings and we could not call them either in C++ API or in
TorchScript. This PR make these two APIs available purely in C++ (with
preserving semantics) and can be used in C++ API and TorchScript

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/24342

Differential Revision: [D16923271](https://our.internmc.facebook.com/intern/diff/D16923271)

wanchaol committed:
[skip ci][WIP][autograd][c++ API] bind autograd functions into C++

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

yf225 commented on torch/csrc/autograd/autograd.cpp:
```diff
@@ -0,0 +1,129 @@
+      at::TensorList tensors,
+      at::TensorList grad_tensors,
+      c10::optional<bool> keep_graph,
+      bool create_graph) {
```
It would be great to preserve the Python equivalent function's signature `torch.autograd.backward(tensors, grad_tensors=None, retain_graph=None, create_graph=False, grad_variables=None)`, so that when a user tries to port code using `torch.autograd.backward` from Python to C++ API it would be easier to do so.

wanchaol replied:
I checked the autograd python signature, `grad_variables` is deprecated and not encouraged to use, user will receive runtime error if they specify `grad_variables`. `retain_graph` is the same as `keep_graph`, as we don't have `NamedType` in C++, the user should not notice any difference?

yf225 replied:
Ignoring `grad_variables` looks good. For `retain_graph`, it seems that if its value is defined, it can be different from `create_graph`? https://github.com/pytorch/pytorch/blob/1df57c943f7bbdec8285c9675dbc3093788b783d/torch/autograd/__init__.py#L88-L89 Would be great to verify this

yf225 commented on torch/csrc/autograd/autograd.cpp:
```diff
@@ -0,0 +1,129 @@
+      at::TensorList grad_outputs,
+      c10::optional<bool> keep_graph,
+      bool create_graph,
+      bool allow_unused) {
```
Ditto for torch.autograd.grad signature: `torch.autograd.grad(outputs, inputs, grad_outputs=None, retain_graph=None, create_graph=False, only_inputs=True, allow_unused=False)`

wanchaol replied:
same for `only_inputs`, it's explicitly deprecated and it's a no-op in python, I think C++ API can just ignore this? Other than the deprecated flags, I think we matched the `python` semantics/signatures in `autograd.h`.

yf225 commented on torch/csrc/autograd/autograd.cpp:
```diff
@@ -0,0 +1,129 @@
+            std::cout<<"grad :" << grads[0] << std::endl;
+        } else {
+            AT_ERROR("element %d of gradients tuple is not a Tensor or None", i);
+        }
```
IIUC part of the code here is based on `_make_grads()` in `torch/autograd/__init__.py` - since we are directly replicating all the logic in that file, it would be great to preserve the original function structure wrt. `backward()`/`grad()` calling `_make_grads()` first and then call `run_backward()`, which makes Python/C++ code parity maintenance easier in the future.

ezyang replied:
Or can we just call the C++ implementation directly from Python?

wanchaol replied:
@ezyang that would make more sense to me on avoid maintaining two equivalent APIs, do you prefer me making it into a separate PR or combine things into this PR?

ezyang replied:
Up to you, I generally like doing it in the same PR, mostly because it prevents us from forgetting to deduplicate later ;)

ezyang replied:
Though, agreeing with Will, it's weird that the logic isn't replicated faithfully here

yf225 replied:
> Or can we just call the C++ implementation directly from Python?

+1 here, if it's not too hard, we should just call the C++ implementation directly from Python.

wanchaol replied:
Yeah I will investigate how hard it is to direct call this API from python. For the reason that the logic isn't faithfully replicate here, it's because the current python API is making None objects first for `grad_outputs` and then `make_grads` construct another variable list with `ones_like` replacing some Nones, then in the THPEngine_run_backward we loop over the list again to get rid of Nones.  C++ API does not have None but only have undefined tensor, so if we fully replicate the logic we will constantly create those undefined tensors for those loops, which might not be efficient, we can just wrap the whole logic in `make_grads` to make it more clear.

yf225 commented:
@ezyang Would you like to review autograd engine usage code here?

wanchaol committed:
Update on "[skip ci][WIP][autograd][c++ API] bind autograd functions into C++"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

wanchaol committed:
Update on "[autograd][c++ API] bind autograd functions into C++"


Summary:

Right now the two APIs that provided in autograd package only have
python bindings and we could not call them either in C++ API or in
TorchScript. This PR make these two APIs available purely in C++ (with
preserving semantics) and can be used in C++ API and TorchScript

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/24342

wanchaol committed:
Update on "[autograd][c++ API] bind autograd functions into C++"


Summary:

Right now the two APIs that provided in autograd package only have
python bindings and we could not call them either in C++ API or in
TorchScript. This PR make these two APIs available purely in C++ (with
preserving semantics) and can be used in C++ API and TorchScript

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/24342

wanchaol committed:
Update on "[autograd][c++ API] bind autograd functions into C++"


Summary:

Right now the two APIs that provided in autograd package only have
python bindings and we could not call them either in C++ API or in
TorchScript. This PR make these two APIs available purely in C++ (with
preserving semantics) and can be used in C++ API and TorchScript

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/24342

ezyang commented on torch/csrc/autograd/autograd.cpp:
```diff
@@ -0,0 +1,138 @@
+#include <torch/csrc/autograd/function.h>
+
+namespace torch {
+namespace autograd {
```
It would be very helpful to say "hey this code duplicates existing logic at X" here, as a reminder for people who are updating the code to update both places

ezyang commented on torch/csrc/autograd/autograd.cpp:
```diff
@@ -0,0 +1,138 @@
+    TORCH_CHECK(
+        gradient_edge.function,
+        "element %d of tensors does not require grad and does not have a grad_fn",
+        i);
```
This is not the correct syntax for `TORCH_CHECK`. `"element ", i, " of tensors" ...`

ezyang reviewed (approved):
Logic looks reasonable

yf225 commented on torch/csrc/autograd/autograd.h:
```diff
@@ -1,17 +1,25 @@
+  TORCH_API void backward(
+      const variable_list& tensors,
+      const variable_list& grad_tensors={},
+      c10::optional<bool> keep_graph=c10::nullopt,
```
nit: would be great to name this `retain_graph` to match the Python API.

wanchaol replied:
I noticed that in `variable.h` the C++ API is using `keep_graph` instead of `retain_graph` [here](https://github.com/pytorch/pytorch/blob/e7a9b0d62f6ca6f3e9f350e827ff7b58a2cb7bdb/torch/csrc/autograd/variable.h#L236), Should we also change this API to make these two aligned?

yf225 replied:
hmm yes I think we should (this can happen in a separate PR if you prefer).

yf225 commented on torch/csrc/autograd/autograd.h:
```diff
@@ -1,17 +1,25 @@
+      const variable_list& outputs,
+      const variable_list& inputs,
+      const variable_list& grad_outputs={},
+      c10::optional<bool> keep_graph=c10::nullopt,
```
nit: same here `keep_graph` -> `retain_graph`

wanchaol committed:
Update on "[autograd][c++ API] bind autograd functions into C++"


Summary:

Right now the two APIs that provided in autograd package only have
python bindings and we could not call them either in C++ API or in
TorchScript. This PR make these two APIs available purely in C++ (with
preserving semantics) and can be used in C++ API and TorchScript

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/24342

wanchaol committed:
Update on "[autograd][c++ API] bind autograd functions into C++"


Summary:

Right now the two APIs that provided in autograd package only have
python bindings and we could not call them either in C++ API or in
TorchScript. This PR make these two APIs available purely in C++ (with
preserving semantics) and can be used in C++ API and TorchScript

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/24342

wanchaol committed:
Update on "[autograd][c++ API] bind autograd functions into C++"


Summary:

Right now the two APIs that provided in autograd package only have
python bindings and we could not call them either in C++ API or in
TorchScript. This PR make these two APIs available purely in C++ (with
preserving semantics) and can be used in C++ API and TorchScript

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/24342

wanchaol committed:
Update on "[autograd][c++ API] bind autograd functions into C++"


Summary:

Right now the two APIs that provided in autograd package only have
python bindings and we could not call them either in C++ API or in
TorchScript. This PR make these two APIs available purely in C++ (with
preserving semantics) and can be used in C++ API and TorchScript

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/24342

hczhu commented on torch/csrc/autograd/autograd.cpp:
```diff
@@ -0,0 +1,145 @@
+  variable_list grad_inputs = Engine::get_default_engine().execute(
+      roots, grad_outputs, keep_graph, create_graph, output_edges);
+  // check if grad_inputs contains None or not base on the allow_unused flag
+  if (inputs.empty()) {
```
Just happened to see this piece of code. It seems not right. `num_inputs` will always be 0.

hczhu commented on torch/csrc/autograd/autograd.cpp:
```diff
@@ -0,0 +1,145 @@
+  // check if grad_inputs contains None or not base on the allow_unused flag
+  if (inputs.empty()) {
+    size_t num_inputs = inputs.size();
+    for (size_t i = 0; i < num_inputs; ++i) {
```
Just happened to see this piece of code. It seems not right. `num_inputs` will always be 0.

wanchaol replied:
yeah good catch! seems we don't have test cases to catch this bug... Let me quickly submit a fix.
